[Documentation] Update instructions to build SWT binaries locally

fedejeanne opened the pull request:
Contributes to #1171 

See https://github.com/eclipse-platform/eclipse.platform.swt/pull/1167#issuecomment-2052499527

Move some sections of the documentation to the general `Readme.md` file and change some outdated/unclear documentation in Windows.

FYI I provided 2 commits for clarity but I will squash them before merging.

HannesWell reviewed (commented):
Thank you @fedejeanne for this.
I have added some comments below how this could be further enhanced.

fedejeanne commented on bundles/org.eclipse.swt/Readme.Win32.md:
```diff
@@ -6,29 +6,24 @@ The SWT windows port is based on the Win32 api.
+  * Any 'bin' project
+* Make sure you imported the projects that reflects your platform _e.g._ `win32`
+* Make sure you imported the fragment that reflects your architecture _e.g._ `org.eclipse.swt.win32.win32.x86_64`
+* In the snippet project, rename the file `.classpath_win32` to `.classpath` and refresh the project.
```
Are you sure?
I imported the stuff on my own into an existing repository (_i.e._ NOT from a fresh "contributor" installation like it's mention in [Contributing.md](https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#creating-an-eclipse-development-environment)) and the dependency the `swt`-project was missing.

HannesWell replied:
> Are you sure? 

I'm quite sure, yes. :) This became obsolete with https://github.com/eclipse-platform/eclipse.platform.swt/pull/1008.
But maybe we are not talking about the same?

> I changed the part about "_Downlad and install Eclipse_" to use the pre-configured IDE mentioned in Contributing.md and moved it to the general `Readme.md`. This simplifies things a lot.

I'll try to have a look at that change tomorrow. It's too late now.

HeikoKlare replied:
Maybe the confusion is that this is about the platform-specific classpath files for the `swt.snippets` project and not for the `swt` project. Only the latter ones have been replaced by central classpath files linked into the SWT fragment projects, haven't they?
There is also still the according Oomph setup task:
https://github.com/eclipse-platform/eclipse.platform.swt/blob/6ebc3e2b57267c7b104a7ab3e2b19d706b6b83b4/bundles/org.eclipse.swt.tools/Oomph/platformSwt.setup#L50-L53

fedejeanne replied:
@HannesWell I changed the text and removed it altogether anyway, so there's no need to rename any files while following the instructions in the (new) section *Running the snippets*, which is now in `Readme.md`

fedejeanne commented on bundles/org.eclipse.swt/Readme.Linux.md:
```diff
@@ -32,7 +20,8 @@ You need to install the following on your system:
-To build SWT natives run the `Build-SWT-native-binaries-for-running-platform` Maven launch configuration.
+See **Building and Testing locally** in [Readme.md](Readme.md)
+
 Alternatively run the `build.sh` script in `binaries/org.eclipse.swt.gtk.linux.${arch}/bin/library` directory. To
```
👍

fedejeanne reviewed (commented):
Thanks for the tips, @HannesWell , I changed some stuff based on them, WDYT? 

I can squash all my commits before merging.

HannesWell reviewed (commented):
I just had another pass over the documentation and since there were many small points to further enhance (mainly in the already existing doc, but it IMHO fits into this change) I just applied the suggestions by myself and pushed them in a separate commit.
Please incoporate them as you see fit.
A short overview is
- Removed outdated information about Hudson builds.
- Remove links and mentioning of specific JDK vendors. From my impression there is not only one source for JDK binaries anymore and I assume that everyone that wants to contribute to SWT has already set up a JDK anyways.
If we want to provide a link, I suggest to use a neutral place like:
https://adoptium.net/marketplace/?version=17
- There is no help for Windows and Mac so I removed its mentioning.

Thank you for these updates. They are definitely due.

HannesWell commented on bundles/org.eclipse.swt/Readme.Win32.md:
```diff
@@ -6,29 +6,24 @@ The SWT windows port is based on the Win32 api.
-* From swt repo, import most projects (except Cocoa/Gtk, root 'swt' project, once)
-* And import the project that reflects your platform.
-* In the snippet project, add the swt project as dependency to launch snippets.
+* Install EGit.
```
Since you have already solved this in a different way, but in general would have been simpler if we would just write that one should import `Eclipse projects` at specific locations, which would already exclude for example the `binaries` project which is just Maven project.

HannesWell commented on bundles/org.eclipse.swt/Readme.Linux.md:
````diff
@@ -32,12 +20,20 @@ You need to install the following on your system:
+or --help to see more options.
+
+The JDK against which the SWT natives are built must be specified in the `SWT_JAVA_HOME` environment variable:
+  ```
+  SWT_JAVA_HOME=<YOUR_JDK_PATH>
+  export SWT_JAVA_HOME
+  ```
````
I'm not so much a Linux export, but from the existing build scripts I think that's the right thing to do.
Nevertheless it would be great if someone with more knowledge could confirm this.

HeikoKlare replied:
Yes, it is correct to set `SWT_JAVA_HOME` (this way).

Still, the instruction do not work as the whole setup relies on being executed in a properly configured binary fragment project and not within the general `org.eclipse.swt` project:
* The default `OUTPUT_DIR` specification in `build.sh` assumes that the `build.sh` is _not_ executed from within the `org.eclipse.swt` project but from without the symlink created for the binaries fragment project, see #1127.
* There are further files required (like `make_common.mak`), which will only be available when properly linked together in a OS-/WS-/Arch-specific fragment project.

So the instruction have to be rewritten to be executed from within a binaries fragment, but then you need to describe as well how to set up that fragment (create the symlink etc, which is automatically done by Eclipse or the Maven build, respectively). Since it is very easy to create the binaries via Maven, I am not sure whether it is worth to describe all that manual setup (and keep it up to date), as that is already done by Maven (and thus documented in the Maven configuration). Whoever does not want to use Maven can just replicate by hand what Maven does (as that's what has to be described in these instructions otherwise).

fedejeanne replied:
I removed all the manual steps in d556545966b1beefd598027b20675dfa34cde7e6.

@HannesWell we can add a permanent link or other further improvements after this PR ;-)

HannesWell commented on bundles/org.eclipse.swt/Readme.macOS.md:
```diff
@@ -82,13 +82,30 @@ More information on how to use the generator is available [here](https://www.ecl
-    mvn clean package -Dnative=cocoa.macosx.x86_64
+You need to install the following on your system:
 
+* TBD
+* Java 17 JDK
```
@HeikoKlare or @Phillipus can you name the tools that have to be installed on a mac in order to be able to build SWT locally for mac?

Phillipus replied:
@HannesWell i just have Java and Maven.

HannesWell replied:
So on Mac everything to compile C code is already installed out of the box?

Phillipus replied:
Good point. I have Apple's `Command Line Tools` installed which is kind of a lightweight version of `Xcode`. So I would add that one of these needs to be installed as a requirement.

HeikoKlare replied:
I can confirm that `Command Line Tools` are sufficient. I am not sure whether it is worth to note that if you want to build the binaries with gcc, you need to install that on MacOS by hand. By default `cc` will execute a `clang` compiler on MacOS.

HeikoKlare commented on bundles/org.eclipse.swt/Readme.Win32.md:
```diff
@@ -2,77 +2,38 @@
   Either select the components "MSVC C++-x64/x86-Buildtools" and "Windows 10 SDK"
   or the workload "Desktop development with C++" which includes the required components and some more.
+* Java 17 JDK
 * (Optional) Install Cygwin
```
I would remove the reference to Cygwin, even when marked as "optional".

HeikoKlare commented on bundles/org.eclipse.swt/Readme.Win32.md:
```diff
@@ -2,77 +2,38 @@
+
+See **Building and Testing locally** in [Readme.md](Readme.md)
+
+Alternatively run the `build.bat` script under `org.eclipse.swt\Eclipse SWT PI\win32\library\` directory.
+To use the locally build natives run the script with the 'install' and `clean` arguments (`.\build.bat clean install`).
```
Does not work for the same reasons as described for Linux (except the `OUTPUT_DIR` problem).

HeikoKlare commented on bundles/org.eclipse.swt/Readme.macOS.md:
```diff
@@ -82,13 +82,30 @@ More information on how to use the generator is available [here](https://www.ecl
+
+See **Building and Testing locally** in [Readme.md](Readme.md)
+
+Alternatively run the `build.sh` script under `org.eclipse.swt\Eclipse SWT PI\cocoa\library\` directory.
+To use the locally build natives run the script with the 'install' and `clean` arguments (`./build.sh clean install`).
```
Does not work, see explanation for Linux.

fedejeanne replied:
Removed in d556545966b1beefd598027b20675dfa34cde7e6

HeikoKlare commented on bundles/org.eclipse.swt/Readme.md:
```diff
@@ -3,12 +3,32 @@ org.eclipse.swt
+
+* **Natives Part**
+
+Make sure the binary project for your platform is imported and open in your workspace,
+_for example_
+  * `binaries/org.eclipse.swt.win32.win32.x86_64` for Windows or
+  * `binaries/org.eclipse.swt.gtk.linux.x86_64` for Linux or
+  * `binaries/org.eclipse.swt.cocoa.macosx.x86_64` for Mac
```
Maybe it is worth to also make some note on `aarch64`, as currently these examples only refer to fragments for environments with x86 architecture but ARM is becoming more and more common.

fedejeanne replied:
Mentioned in d556545966b1beefd598027b20675dfa34cde7e6

HeikoKlare reviewed (commented):
If I am not mistaken, the descriptions how to build the binaries by hand (directly executing the `build.sh` / `build.bat` scripts), do not work as described. I would propose to completely remove these descriptions, as there is a Maven configuration that does everything required and replication all the information in a documentation leads to unnecessary duplications that needs to be kept consistent.

HannesWell commented:
> Since it is very easy to create the binaries via Maven, I am not sure whether it is worth to describe all that manual setup (and keep it up to date), as that is already done by Maven (and thus documented in the Maven configuration). 

Just as a quick-reply for now: 
Yes, I agree. But to make it simpler to find the relevant part in the pom I suggest to also add a perma-link to the configuration of the corresponding antrun-mojo configuration.
Of course this might also change in the future, but if it is forgotten to update the link (which actually can be done after the fact since it contains the commit Id :/), then it is at least easier to find to up-to-date info.

fedejeanne commented:
@HannesWell , @HeikoKlare , thank you for the comments, I addressed almost all of them in d556545966b1beefd598027b20675dfa34cde7e6 and made some other minor changes. I am also in favor of simplifying these instructions as much as possible and let the "hackers" figure out the alternative ways (in case they don't want to run Maven) on their own.

I'm positive that the current state of these instructions is good enough to compile the binaries in the majority of the setups.

Once this PR is good to go, I will squash the commits.

@HannesWell We can add a permanent link to the desired commit after merging this PR if you want to. Just to be sure: you want to link this line (but not in master), right? 

https://github.com/eclipse-platform/eclipse.platform.swt/blob/979d3f1580b4786aee1efe046da8a6f836300804/binaries/pom.xml#L93

How would you phrase it in the documentation (in the follow-up PR)?

fedejeanne commented:
The failing tests are unrelated

SyntevoAlex commented:
Sorry, but inverse changes are actually needed.

You are removing steps for manual building, but that's exactly what I was always doing, and I was pretty much committer number one for a few years.

I was running steps manually because I'm not using Eclipse for developing SWT. And when I run maven from IDEA, it downloads unneeded stuff (tycho etc) for some 20 minutes before actually building.

HeikoKlare commented:
> You are removing steps for manual building, but that's exactly what I was always doing, and I was pretty much committer number one for a few years.

There seems to be some confusion here, as the proposed change removes nothing but redundant and outdated documentation. You can still do a "manual build" (i.e., do by hand what Maven does automatically). Having an explanation of that process in the documentation would be nothing else than writing down the according Maven step in natural language:
https://github.com/eclipse-platform/eclipse.platform.swt/blob/979d3f1580b4786aee1efe046da8a6f836300804/binaries/pom.xml#L92-L134

So as @HannesWell proposed (https://github.com/eclipse-platform/eclipse.platform.swt/pull/1174#issuecomment-2068594910), adding a link to the Maven configuration in the documentation for those who want to perform the steps by hand would be good.

SyntevoAlex commented:
I still think that instead of removing documentation, it should be fixed, so removing is half step in wrong direction. It is not redundant, because parsing maven files is hard for people who never dealt with maven before (such as me).

I'm merely commenting and do not intend to hold this PR.

merks commented:
I think the point is that while the actual logic as spelled out in the pom.xml may evolve in the future (logic that is well tested by the builds and by the other developers) the documentation will forever need to evolve manually in lockstep.  That's likely to be forgotten or not to happen.  I image that most people will be happy if they can invoke Maven without further thought about the details, and even if that takes 20 minutes the first time to fill the cache, that's probably for most people better tradeoff than trying to read and to follow more detailed documentation instructions.

So I do understand your concerns, I think this is a little like with Oomph where it's must easier to maintain something that's regularly and automatically used, than it is to maintain documentation that's rarely read, rarely used, and error prone to follow exactly to the letter...

fedejeanne commented:
@SyntevoAlex I agree with @HannesWell , @HeikoKlare and @merks here: pointing to the Maven goal that does the trick (in a follow-up PR) is enough here so I wouldn't hold this PR just to add the "manual" steps which are likely to be outdated in the near future anyway.

So as soon as @HeikoKlare and @HannesWell give me the green light (I hope I addressed every one of your comments), I'm squashing + merging :-)

HeikoKlare commented on bundles/org.eclipse.swt/Readme.Linux.md:
````diff
@@ -4,40 +4,29 @@ The SWT Linux port is based on the GTK.
-use the locally build natives run the build script with the 'install' argument
-(`./build.sh install`) or --help to see more options.
+The JDK against which the SWT natives are built must be specified in the `SWT_JAVA_HOME` environment variable:
+  ```
+  export SWT_JAVA_HOME=<YOUR_JDK_PATH>
+  ```
````
Setting `SWT_JAVA_HOME` should not be necessary when using the Maven build.

fedejeanne replied:
Removed it from all 3 platforms in 74bda8a9bb2fd3618d2c3b98c41fb22fe9074947

HeikoKlare commented:
> I still think that instead of removing documentation, it should be fixed, so removing is half step in wrong direction. It is not redundant, because parsing maven files is hard for people who never dealt with maven before (such as me).

I understand your point that it is harder to parse Maven files than reading documentation that explains the Maven files in comparably simple language (even though that does not change anything w.r.t. it being redundant). However, maintaining the documentation is even harder, as only very few people will need that documentation (as stated by @merks). In my opinion, it even leads to the opposite: the people that may be interested in compiling the natives will run across all the information on how to execute the build.sh/build.bat and be tempted to follow that approach instead of simply executing the Maven build. So the documentation would have to be structured very well to lead the people in the proper direction, i.e., to just use the simple, Maven-based solution and understand that all the information about the manual build is irrelevant for them.

fedejeanne commented:
Addressed https://github.com/eclipse-platform/eclipse.platform.swt/pull/1174#discussion_r1576022056 and also added the perma-link from https://github.com/eclipse-platform/eclipse.platform.swt/pull/1174#issuecomment-2071705795 in 74bda8a9bb2fd3618d2c3b98c41fb22fe9074947

HannesWell reviewed (commented):
Hoping that it will be a good compromise I made more suggestions for improvements below.
